RPIPs/RPIP-47.md

```diff
@@ -26,34 +26,35 @@ Additionally, the pDAO must have the ability to effect change in ways that benef
 This RPIP is part of a set of proposals motivated by a desire to rework Rocket Pool's tokenomics to ensure the protocol’s continued value, development, and longevity. For more details, see the supporting documentation [here](../tokenomics-explainers/001-why-rework). 
 
 ## Specification
-### Megapool delegates
-- `Megapool delegate` contracts SHALL have an expiration block (ie, execution layer block)
-- When a new `megapool delegate` is released, the contract upgrade SHALL include:
-  - Setting the new `megapool delegate`'s expiration block to "no expiration"
-  - Setting the previous `megapool delegate`'s expiration block to occur `delegate_upgrade_buffer` after the upgrade
-- A node operator SHALL be able to upgrade their active `megapool delegate` to the latest `megapool delegate` at any time (including after the active `megapool delegate`'s expiration block)
-- After the active `megapool delegate`'s expiration block:
-  - Functions in the active megapool delegate SHALL revert when not called by a protocol contract
+
+### Upgrade Buffer Enforcement
+- Each `Megapool delegate` contract SHALL have an execution layer `expiration_block` variable which should initialize to "no expiration"
+- Each `Megapool delegate` contract SHALL have a non-modifiable `upgrade_buffer` value
+  - The `Megapool delegate`'s `upgrade_buffer` SHALL be 120 days
+- The only permitted modification to a `Megapool delegate`'s `expiration_block` variable SHALL be:
+  - Any protocol contract may set its value to now + its `upgrade_buffer`
+  - Any protocol contract may set its value to "no expiration"
```

Comment:
what's the use case here?

Reply:
Lol Knoshua asked the same thing and I resolved it. It's mainly intended as a fallback. If we discover something is wrong that can't be quickly fixed, it might make more sense to set it back to no-expire rather than pushing it 120 days.

Reply:
I don't think I like that at all. We could just redeploy a copy of the old thing as latest. Then it would be "no expiration" and also properly work for "update to latest". I recommend removing this.

Reply:
I don't think I agree. Redeploying a copy is more work and forces someone to pay gas to move people from one identical contract to another. The contracts need to support no-expiration status anyway as the initial value. Can you elaborate on why you don't want this?

Reply:
The example you gave is:

- GoodDelegateA has been deployed a while
- BadDelegateB is deployed; GoodDelegateA gets an expiration block
- We realize and say "oops". We set GoodDelegateA's expiration block to nothing

In this new state we have:
- Two non-expiring delegates that people might be on
- No way for people on BadDelegateB to get to a GoodDelegate
- A trap -- anyone that upgrades ends up on a BadDelegate
- Another trap -- anyone that makes a new node will use the latest... which is BadDelegateB

-----

My proposal would instead be to deploy GoodDelegateC (which may be a copy of A or not; but let's say it is for now).

In that state we have:
- a single non-expiring delegate (C)
- a way for people on BadDelegate to get to a GoodDelegate
- no traps

Yes, people may  need to waste gas going from A to C, which is the same. They only need to do this if (A) A is expiring, (B) the latest delegate isn't C (cuz being force upgraded to a copy isn't at all damaging), (C) they're not wanting to upgrade to the latest delegate (ditto)

Reply:
Discussed some with Val on discord. TL;DR is that if we were to keep it, it would be to handle unforeseen issues that we can't think of. Any feasible issue should be resolvable by just redeploying a known-good delegate as the latest, which is slightly more painful. In most cases we can think of, you'd need to replace the delegate anyway to prevent voluntary upgrades to a bad contract.

Reply:
So line removed and comment resolved.